### PR TITLE
build providers: move to buildd images

### PR DESCRIPTION
**Pull request by sergiusens**
This enables support for all architectures and core20 for LXD

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/build_providers/_lxd/_images.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def get_image_source(*, base: str) -> Dict[str, str]:
     return dict(
         type="image",
         mode="pull",
-        server="https://cloud-images.ubuntu.com/minimal/releases/",
+        server="https://cloud-images.ubuntu.com/buildd/daily",
```
Comment (cjp256) [inline snapcraft/internal/build_providers/_lxd/_images.py]: would this affect any images today (negatively)?

File: snapcraft/internal/build_providers/_lxd/_lxd.py [hunk 2]
```diff
@@ -365,6 +382,49 @@ def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
...
         self._run(command=["/bin/bash"])
 
+    def _wait_for_network(self) -> None:
+        while True:
+            if self._container is None:
+                raise RuntimeError("Attempted to use container before starting.")
+            if any(
+                a["address"]
+                for a in self._container.state().network["eth0"]["addresses"]
+                if a["family"] == "inet"
```
Comment (cjp256) [inline snapcraft/internal/build_providers/_lxd/_lxd.py]: Is ipv6 supported today?  or will lxd tunnel 4to6?

File: snapcraft/internal/build_providers/_lxd/_lxd.py [hunk 3]
```diff
@@ -365,6 +382,49 @@ def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
...
+            ):
+                break
+            sleep(1)
```
Comment (cjp256) [inline snapcraft/internal/build_providers/_lxd/_lxd.py]: Maybe make 0.1?  I don't think that'd be unreasonable - especially if it saves me a few seconds each day :stuck_out_tongue_winking_eye:

File: snapcraft/internal/build_providers/_lxd/_images.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/internal/build_providers/_lxd/_images.py]: Yes, they will have missing packages in some cases, but will match closely what is on launchpad. This should be fine considering that we had the "preview" message when using "--use-lxd".

File: snapcraft/internal/build_providers/_lxd/_lxd.py (hunk 3 above)
Comment (sergiusens) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: sure

File: snapcraft/internal/build_providers/_lxd/_lxd.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: there's always an ipv6 address assigned, so this is not the best mechanism to have if the network is ipv6 only and we might need to fallback to connectivity checks (but that would make all builds require being online). Not sure what is best at this point with this. So far, for snapcraft we have been recommending lxd defaults, which setup ipv4

Comment (sergiusens) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: @stgraber  or @Saviq do you have suggestions for this?

Comment (stgraber) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: LXD sets up a private IPv4 and IPv6 subnet by default.
As we use a private IPv6 subnet, the C library will prefer IPv4 over it, however should connecting over IPv4 fail, IPv6 will be attempted.

Comment (cjp256) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: Should we wait for both? Or just get started once ipv4 is configured?

File: snapcraft/internal/build_providers/_lxd/_images.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/internal/build_providers/_lxd/_images.py]: roger

File: snapcraft/internal/build_providers/_lxd/_lxd.py (hunk 2 above)
Comment (stgraber) [reply snapcraft/internal/build_providers/_lxd/_lxd.py]: The normal recommend is to wait for what you care about. In most cases you don't care about an address, you care about being able to connect to stuff.

In our own CI, our go-to test is to run `getent hosts linuxcontainers.org` in the container. When that succeeds it means we have an IP and DNS working, doesn't matter if over v4 or v6.

**Comment by sergiusens** [comment]
On jue, 2020-03-26 at 10:52:48 -0300, Stéphane Graber wrote:

> @stgraber commented on this pull request.
>
>
>
>> @@ -365,6 +382,49 @@ def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
>      def shell(self) -> None:
>          self._run(command=["/bin/bash"])
>  
> +    def _wait_for_network(self) -> None:
> +        while True:
> +            if self._container is None:
> +                raise RuntimeError("Attempted to use container before starting.")
> +            if any(
> +                a["address"]
> +                for a in self._container.state().network["eth0"]["addresses"]
> +                if a["family"] == "inet"
>
> The normal recommend is to wait for what you care about. In most cases you don't care about an address, you care about being able to connect to stuff.
>
> In our own CI, our go-to test is to run `getent hosts linuxcontainers.org` in the container. When that succeeds it means we have an IP and DNS working, doesn't matter if over v4 or v6.

True, I was trying to think of the repeatable I-am-offline case and thinking
about this a bit more, we really don't need an IP at all at that point.

-- 
---
Sergio Schvezov

File: snapcraft/internal/build_providers/_lxd/_lxd.py [hunk 4]
```diff
@@ -361,6 +378,91 @@ def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
     def shell(self) -> None:
         self._run(command=["/bin/bash"])
 
+    def _wait_for_network(self) -> None:
+        for i in range(5):
```
Comment (cjp256) [inline snapcraft/internal/build_providers/_lxd/_lxd.py:382]: i think a longer timeout is desirable. this would be 2.5 secs?

**Pull request merged**